lib/cloudware/models/deployment.rb

```diff
@@ -0,0 +1,99 @@
+# frozen_string_literal: true
+
+require 'models/concerns/provider_client'
+require 'models/application'
+require 'models/machine'
+require 'models/context'
+require 'pathname'
+
+require 'erb'
+
+module Cloudware
+  module Models
+    class Deployment < Application
+      include Concerns::ProviderClient
+
+      SAVE_ATTR = [:template_path, :name, :results, :replacements].freeze
+      attr_accessor(*SAVE_ATTR, :context)
+
+      define_model_callbacks :deploy
+
+      before_deploy :validate_replacement_tags
+      before_deploy :validate_context
+      before_deploy :validate_no_existing_deployment
+
+      def template
+        return raw_template unless replacements
+        replacements.reduce(raw_template) do |memo, (key, value)|
+          memo.gsub("%#{key}%", value.to_s)
+        end
+      end
+
+      def deploy
+        run_callbacks(:deploy) do
+          if errors.blank?
+            run_deploy
+          else
+            msg = ERB.new(<<-TEMPLATE, nil, '-').result(binding).chomp
+Failed to deploy resources. The following errors have occurred:
+<% errors.messages.map do |key, messages| -%>
+<% messages.each do |message| -%>
+<%= key %>: <%= message %>
+<% end -%>
+<% end -%>
+TEMPLATE
+            raise ModelValidationError, msg
+          end
+        end
+      end
+
+      def destroy
+        context&.remove_deployment(self)
+        provider_client.destroy(tag)
+      end
+
+      def machines
+        Machine.build_from_context(self)
```

Comment:
As mentioned above, this is a confusing name.  Perhaps: `Machine.all(self)`

Also, as it's an operation on the context, perhaps the content of `Machine.build_from_context` would be better living here?